### PR TITLE
decouple pyo3-ffi-check from MSRV

### PR DESCRIPTION
**Pull request by davidhewitt**
Inspired by #3276, I figured we should perhaps just exclude pyo3-ffi-check from the workspace during MSRV jobs. It's only an internal tool, we already don't run it on MSRV, and hopefully dependabot version bumps for its deps can then be trivially merged.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -57,8 +57,11 @@ jobs:
         run: echo LD_LIBRARY_PATH=${pythonLocation}/lib >> $GITHUB_ENV
 
       - if: inputs.msrv == 'MSRV'
-        name: Prepare minimal package versions (MSRV only)
-        run: nox -s set-minimal-package-versions
+        name: Prepare package versions for MSRV
+        run: |
+          sed -i '/pyo3-ffi-check/d' Cargo.toml  # we allow pyo3-ffi-check to use newer packages, as an internal tool
+          git diff Cargo.toml
+          nox -s set-minimal-package-versions
```
Comment (alex) [inline .github/workflows/build.yml:64]: Can we just run the various commands with `--workspace --exclude pyo3-ffi-check`?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -57,8 +57,11 @@ jobs:
         run: echo LD_LIBRARY_PATH=${pythonLocation}/lib >> $GITHUB_ENV
 
       - if: inputs.msrv == 'MSRV'
-        name: Prepare minimal package versions (MSRV only)
-        run: nox -s set-minimal-package-versions
+        name: Prepare package versions for MSRV
+        run: |
+          sed -i '/pyo3-ffi-check/d' Cargo.toml  # we allow pyo3-ffi-check to use newer packages, as an internal tool
```
Comment (adamreichold) [inline .github/workflows/build.yml:62]: I think the Noxfile already mentions `pyo3-ffi-check` by name, excluding it from linting.

So I see two alternatives to this "open-coded" approach:
* We move this change into the `set-minimal-package-versions` session to concentrate this in one place.
* We fully remove pyo3-ffi-check from the workspace, e.g. by adding an empty `[workspace]` to its `Cargo.toml`, and build it only when needed.

Personally, I would prefer the second option as the cleaner one.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -51,6 +51,11 @@ jobs:
           key: check-msrv-1.56.0
         continue-on-error: true
       - run: python -m pip install --upgrade pip && pip install nox
+      - name: Prepare package versions for MSRV
+        run: |
+          sed -i '/pyo3-ffi-check/d' Cargo.toml  # we allow pyo3-ffi-check to use newer packages, as an internal tool
+          git diff Cargo.toml
+          nox -s set-minimal-package-versions
       - name: Prepare minimal package versions
         run: nox -s set-minimal-package-versions
```
Comment (adamreichold) [inline .github/workflows/ci.yml:60]: This is now called twice by the CI job.

Comment (davidhewitt) [reply .github/workflows/ci.yml]: Thanks, fixed.

File: .github/workflows/build.yml (hunk 2 above)
Comment (davidhewitt) [reply .github/workflows/build.yml]: That's probably cleaner; I've force-pushed to reimplement as the second option.

File: noxfile.py [hunk 4]
```diff
@@ -534,8 +526,12 @@ def load_pkg_versions():
 
 @nox.session(name="ffi-check")
 def ffi_check(session: nox.Session):
-    session.run("cargo", "doc", "-p", "pyo3-ffi", "--no-deps", external=True)
-    _run(session, "cargo", "run", "-p", "pyo3-ffi-check", external=True)
+    path = ["--manifest-path", "pyo3-ffi-check/Cargo.toml"]
+    session.run("cargo", "doc", *path, "-p", "pyo3-ffi", "--no-deps", external=True)
+    _run(
+        session, "cargo", "clippy", "--workspace", "--all-targets", *path, external=True
```
Comment (adamreichold) [inline noxfile.py:532]: This one I do not understand. This is to make sure the ffi-check binary is Clippy clean? But why the `--workspace` flag then?

Comment (davidhewitt) [reply noxfile.py]: Yes, to check the ffi-check binary, and then `--workspace` is to also check `pyo3-ffi-check/macro`. I've force-pushed to add "macro" subdirectory to the workspace section in `pyo3-ffi-check/Cargo.toml`.

Also, good point about rustfmt - I've added that to the `fmt-rust` job.

**Comment by davidhewitt** [comment]
I've pushed a second commit with a tiny tidy-up to the `noxfile` with a `_run_cargo` helper function.

File: noxfile.py [hunk 5]
```diff
@@ -139,18 +133,8 @@ def _check(env: Dict[str, str]) -> None:
...
-                    *feature_set,
-                    "--all-targets",
-                    "--workspace",
-                    # linting pyo3-ffi-check requires docs to have been built or
-                    # the macros will error; doesn't seem worth it on CI
-                    "--exclude=pyo3-ffi-check",
-                    external=True,
-                    env=env,
+                _run_cargo(
+                    session, *feature_set, "--all-targets", "--workspace", env=env
```
Comment (alex) [inline noxfile.py:137]: Missing "check"

**Review by adamreichold — APPROVED** [review]

File: noxfile.py (hunk 5 above)
Comment (davidhewitt) [reply noxfile.py]: Thanks, the classic editing-code-last-thing-at-night mistake - at least we have CI to check me! 😄

**Pull request merged**